src/States.cpp

```diff
@@ -0,0 +1,404 @@
+      throw std::invalid_argument("Length of times must match number of positions"); 
+    }
+    
+    ale::Vec3d velocities = {0.0, 0.0, 0.0};
+    for (ale::Vec3d position : positions) {
+      m_states.push_back(ale::State(position, velocities)); 
+    }
+    m_minimizedCache=false;
+  }
+
+
+  States::States(std::vector<double> ephemTimes, std::vector<ale::Vec3d> positions, 
+                 std::vector<ale::Vec3d> velocities, int refFrame) : 
+    m_ephemTimes(ephemTimes), m_refFrame(refFrame) {
+    
+    if ((positions.size() != ephemTimes.size())||(ephemTimes.size() != velocities.size())) {
+      throw std::invalid_argument("Length of times must match number of positions and velocities."); 
+    }
+    
+    for (int i=0; i < positions.size() ;i++) {
+      m_states.push_back(ale::State(positions[i], velocities[i])); 
+    }
+    m_minimizedCache=false;
+  }
+
+
+  States::States(std::vector<double> ephemTimes, std::vector<ale::State> states, int refFrame) :
+  m_ephemTimes(ephemTimes), m_states(states), m_refFrame(refFrame), m_minimizedCache(false) {
+    if (states.size() != ephemTimes.size()) {
+      throw std::invalid_argument("Length of times must match number of states."); 
+    }
+  }
+
+  // Default Destructor 
+  States::~States() {}
+
+  // Getters 
+  std::vector<ale::State> States::getStates() const {
+    return m_states; 
+  }
+
+  std::vector<ale::Vec3d> States::getPositions() const {
+    // extract positions from state vector
+    std::vector<ale::Vec3d> positions; 
+    
+    for(ale::State state : m_states) {
+        positions.push_back(state.position);
+    }
+    return positions; 
+  }
+
+
+  std::vector<ale::Vec3d> States::getVelocities() const {
+    // extract velocities from state vector
+    std::vector<ale::Vec3d> velocities; 
+    
+    for(ale::State state : m_states) {
+        velocities.push_back(state.velocity);
+    }
+    return velocities; 
+  }
+
+
+  std::vector<double> States::getTimes() const {
+    return m_ephemTimes; 
+  }
+
+
+  int States::getReferenceFrame() const {
+    return m_refFrame; 
+  }
+
+
+  bool States::hasMinimizedCache() const {
+    return m_minimizedCache;
+  }
+
+
+  bool States::hasVelocity() const {
+    std::vector<ale::Vec3d> velocities = getVelocities(); 
+    bool allZero = std::all_of(velocities.begin(), velocities.end(), [](ale::Vec3d vec) 
+                               { return vec.x==0.0 && vec.y==0.0 && vec.z==0.0; });
+    return !allZero;
+  }
+
+
+  ale::State States::getState(double time, ale::interpolation interp) const {
+    std::vector<double> xs, ys, zs, vxs, vys, vzs; 
+    for (ale::State state : m_states) {
+      xs.push_back(state.position.x);
+      ys.push_back(state.position.y);
+      zs.push_back(state.position.z);
+      vxs.push_back(state.velocity.x);
+      vys.push_back(state.velocity.y);
+      vzs.push_back(state.velocity.z);
+    }
+
+    ale::Vec3d position, velocity; 
+
+    if (!hasMinimizedCache()) {
```

Comment:
This should be a switch statement that uses the passed in interpolation type. If linear is passed in then use the linear GSL interpolation. If spline is passed in then use the hermite spline interpolation.

Reply:
@jessemapel Should we let the user ask for a linear interpolation if the cache has been reduced to within some tolerance specifically calculated for the hermite cache? Couldn't that result in worse-than-intended values?